[BEAM-252] Make Regex Transform

eljefe6a opened the pull request:
Jira issue BEAM-252 Make Regex Transform. Adding a pre-built transform for Regex.

kennknowles commented on sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/RegexTransformTest.java:
```diff
@@ -0,0 +1,202 @@
+  }
+
+  @Test
+  public void testReplaceFirstMixed() {
```
For tests that run a pipeline, now they need to be annotated with `@NeedsRunner`, since there is no runner in the core SDK.

eljefe6a commented:
@kennknowles All passing now.

eljefe6a commented:
@kennknowles Anything you need from on this PR?

kennknowles commented:
Nope, just a bit backlogged. Thanks for the ping!

dhalperi commented:
[ignore me, metadata only] R: @kennknowles

eljefe6a commented:
@kennknowles I've updated the PR to the latest `DoFN` interface. Should make for an easier merge.

eljefe6a commented:
@kennknowles I was looking at the WordCount code and realized that should be in `RegexTransform`. I added a `split` method that will allow the word splitting portion of WordCount to be done as `words.apply(RegexTransform.split("\W*"));`.

kennknowles commented:
@eljefe6a indeed, very appropriate. This is still on my radar FYI.

eljefe6a commented:
@kennknowles Are you ok with me merging this?

kennknowles commented:
Actually, I had some more feedback for tightening up the API, and was working on a pull-request-for-your-pull-request. We've been considering moving non-core transforms to extensions. I'll put something together.

Sorry this didn't get prompt attention, but we really need to stick to waiting for an LGTM comment or GitHub "approval".

eljefe6a commented:
I thought this was approved. My mistake.

dhalperi commented:
The other thing I would add is that for very old branches the precommits need to be run.

Specifically, if you're the author: rebase and push so that Jenkins re-runs.

If not, it's trickier -- the merging committer can run the precommit manually (e.g., `mvn clean verify -Prelease`) before merging.

This merge immediately broke the build. I think Kenn has pushed fixes, but we should prevent these build breaks.
